rust/lance-io/src/object_store/credential_vending.rs

```diff
@@ -0,0 +1,878 @@
+
+/// Trait for providing credentials with expiration tracking
+///
+/// Implementations can fetch credentials from various sources (namespace servers,
+/// credential vending machines, etc.) and are usable from Python/Java via FFI.
+#[async_trait]
+pub trait CredentialVendor: Send + Sync {
+    /// Fetch fresh credentials
+    ///
+    /// Returns a tuple of (storage_options, expires_at_millis)
+    /// where storage_options is a map of credential key-value pairs
+    /// and expires_at_millis is the epoch time in milliseconds when credentials expire
+    async fn get_credentials(&self) -> Result<(HashMap<String, String>, u64)>;
+}
+
+/// CredentialVendor implementation that fetches credentials from a LanceNamespace
+pub struct LanceNamespaceCredentialVendor {
+    namespace: Arc<dyn LanceNamespace>,
+    table_id: Vec<String>,
+}
+
+impl LanceNamespaceCredentialVendor {
+    /// Create a new LanceNamespaceCredentialVendor
+    ///
+    /// # Arguments
+    /// * `namespace` - The namespace implementation to fetch credentials from
+    /// * `table_id` - The table identifier
+    pub fn new(namespace: Arc<dyn LanceNamespace>, table_id: Vec<String>) -> Self {
+        Self {
+            namespace,
+            table_id,
+        }
+    }
+}
+
+#[async_trait]
+impl CredentialVendor for LanceNamespaceCredentialVendor {
+    async fn get_credentials(&self) -> Result<(HashMap<String, String>, u64)> {
+        use lance_namespace::models::DescribeTableRequest;
+
+        let request = DescribeTableRequest {
+            id: Some(self.table_id.clone()),
+            version: None,
+        };
+
+        let response = self
+            .namespace
+            .describe_table(request)
+            .await
+            .map_err(|e| Error::IO {
+                source: Box::new(std::io::Error::other(format!(
+                    "Failed to fetch credentials: {}",
+                    e
+                ))),
+                location: location!(),
+            })?;
+
+        let storage_options = response.storage_options.ok_or_else(|| Error::IO {
+            source: Box::new(std::io::Error::other(
+                "storage_options not found in describe_table response",
+            )),
+            location: location!(),
+        })?;
+
+        let expires_at_millis = storage_options
+            .get("expires_at_millis")
+            .and_then(|s| s.parse::<u64>().ok())
+            .ok_or_else(|| Error::IO {
+                source: Box::new(std::io::Error::other(
+                    "expires_at_millis is required in storage_options",
+                )),
+                location: location!(),
+            })?;
+
+        Ok((storage_options, expires_at_millis))
+    }
+}
+
+/// Configuration parameters for credential vending
+#[derive(Debug, Clone)]
+pub struct CredentialVendingParams {
+    /// How early to refresh credentials before expiration (in milliseconds)
+    /// Default: 300,000 (5 minutes)
+    pub refresh_lead_time_ms: u64,
```

Comment:
Similarly, can we use a proper type here like `std::time::Duration`?
```suggestion
    /// How early to refresh credentials before expiration (in milliseconds)
    /// Default: 300,000 (5 minutes)
    pub refresh_lead_time_ms: Duration,
```